File: Graph/310. Minimum Height Trees/minimumHeightTrees.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {number[]}
 */
 var findMinHeightTrees = function(n, edges) {
    //edge case
    if (n === 1) return [0];
    
    let output = [];
    
    // set up map to present graph
    // set up degree for each node
    let map = new Map();
    let degree = new Array(n).fill(0);
    
    for (let i = 0; i < n; i++) {
        map.set(i, []);
    }
    for (let [a, b] of edges) {
        map.get(a).push(b);
        map.get(b).push(a);
        degree[a]++;
        degree[b]++;
    }
    
    // console.log(map, degree)
    
    // record the node with least edges
    let queue = [];
    for (let i = 0; i < degree.length; i++) {
        if (degree[i] === 1) 
            queue.push(i)
    }
    
    console.log(queue)
    while (queue.length) {
        // temp list to save the final results
        let list = [];
        let size = queue.length;
        for (let i = 0; i < size; i++) {
            let current = queue.shift();
            list.push(current);
            // console.log(list, queue)
            for (let n of map.get(current)) {
                degree[n]--;
                if (degree[n] === 1) queue.push(n);
            }
        }
        output = list;
    }
    
    return output;
};

File: Graph/310. Minimum Height Trees/minimumHeightTrees.py
class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        # edge case
        if n == 1:
            return [0]
        
        # record graph and degree
        outupt = []
        map = { i: [] for i in range(n) }
        degree = [0 for _ in range(n) ]
        
        for a, b in edges:
            map[a].append(b)
            map[b].append(a)
            degree[a] += 1
            degree[b] += 1
        
        print(map, degree)
        
        queue = [ i for i, e in enumerate(degree) if e == 1]
        #print(queue)
        
        # traverse using size of queue
        # queue only contains degree of 1
        # refresh list 
        while len(queue):
            list = []
            size = len(queue)
            for _ in range(size):
                current = queue.pop(0)
                list.append(current)
                for n in map[current]:
                    degree[n] -= 1
                    if degree[n] == 1:
                        queue.append(n)
            outupt = list
            
        return outupt